pkg/dm/message_impl.go

```diff
@@ -0,0 +1,83 @@
+package dm
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pingcap/errors"
+
+	"github.com/hanfei1991/microcosm/pkg/p2p"
+)
+
+type MessageIDAllocator struct {
+	mu sync.Mutex
+	id uint64
+}
+
+func (a *MessageIDAllocator) Alloc() uint64 {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.id++
+	return a.id
+}
+
+type Sender interface {
+	SendMessage(ctx context.Context, topic p2p.Topic, message interface{}, nonblocking bool) error
+}
+
+// MessageImpl implement a simple synchronous request/response message pattern since the lib currently only support asynchronous message.
+// Caller should persist the request message if needed.
+// Caller should add retry mechanism if needed.
+type MessageImpl struct {
+	// messageID -> response channel
+	// TODO: limit the MaxPendingMessageCount if needed.
+	pendings    sync.Map
+	idAllocator *MessageIDAllocator
+}
+
+func NewMessageImpl() *MessageImpl {
+	return &MessageImpl{
+		idAllocator: &MessageIDAllocator{},
+	}
+}
+
+// SendMessage sends a message asynchronously.
+func (m *MessageImpl) SendMessage(ctx context.Context, topic p2p.Topic, message Message, sender Sender) error {
```

Comment:
maybe this method is not needed, we have too many `SendMessage` now

Reply:
We need sends a message asynchronously, caller(message_agent) needs to decide for itself whether a message is asynchronous or not.

Reply:
I guess your meaning is when dealing with messages, we should always construct a MessagePair, and use it to synchronously or asynchronously send messages by `SendMessage` or `SendRequest`. 

Both `SendMessage` or `SendRequest` need the actual message and a sender as parameters, for `SendMessage` caller can directly use sender to send the message, no need to construct a MessagePair. This is also stated by the definition of MessagePair: "MessagePair implement a simple synchronous request/response message pattern". And the caller (I guess) will not dynamically choose synchronously or asynchronously sending, most method should only need one of them by the semantic.

For these reason I think wrap a `SendMessage` to MessagePair is not needed.

Reply:
removed